rfcs/0154-Migrate-taskcluster-to-postgres.md

```diff
@@ -0,0 +1,122 @@
+   Azure.
+2. Go fully into normalized SQL tables.
+
+In both steps, Taskcluster will get confidence in the correctness and
+performance of the implementation by first testing it in the development and
+staging environments before running this on the production environment.
+
+Upon completion of step 1, deployers of Taskcluster will need to provide a DB
+URL (maybe several) then run a script to "load" data from Azure to Postgres
+during a downtime.
+
+There will be a version of Taskcluster that supports Postgres and not Azure,
+with the previous version supporting Azure and not Postgres. So upgrading to
+that version is something that happens in a downtime.
+
+## Schema Migration
+
+Taskcluster will handle schema migrations and will have
+all migrations tested with sample data to ensure we get the expected schema. In
+addition, there will be one schema for the whole monorepo, so that schema
+updates happen atomically across all services.
+
+## Backups and Restores
+
+Teams operating Taskcluster will rely on the cloud provider's backup system to
+handle backups and restores.
+
+## DB Version Upgrades Compatibility
+
+When deploying a new version of Taskcluster, any necessary database upgrades
+*must* be applied before any services are upgraded.  Taskcluster will maintain
+the invariant that services expecting database version S can interoperate with a
+database at version V as long as V >= S.
+
+This invariant is maintained through careful attention to the definitions of the
+stored procedures (user-defined functions). These stored procedures are changed
+as necessary during the upgrade process, such that a procedure operates
+identically before and after the ugprade: identical arguments, identical return
+type, and correct behavior.
+
+_Example 1:_ If an upgrade factors a single table into two tables, then a
+procedure which previously queried from the single table would be rewritten
+during the ugprade to perform a join over the two new tables, returning
+appropriate results in the same format as before the upgrade.
+
+_Example 2:_ If you have a query that returns column foo, but in the next
+version your code wants to know the value of column bar as well, then a new
+stored procedure will need to be introduced to return both columns and stop
+using the existing stored procedure that returned the single column. That new
+stored procedure is then deployed before the code that uses it is deployed.
+
+A consequence of this design is that "procedures are forever" -- an upgrade can
+never delete a stored procedure. At worst, when a feature is removed, a stored
+procedure can be rewritten to return an empty result or perform no action.
+
+## Ad-hoc Queries
+
+Direct SQL access to the database is *not allowed*. Taskcluster will allow
+ad-hoc queries on the data-set via stored procedures with access controlled by
+Postgres permissions. This feature will most likely be done after step 2 of the
+transition.
```

Comment:
Will users with sufficient permissions be able to create their own queries, or just reuse stored ones? Will there be restrictions to what a query can write or modify?

This looks like its intent is "you can read certain stuff from the DB through Taskcluster if you have the permissions", which I agree is a great idea. However, the way it's currently worded doesn't seem to rule out the possibility of a user making changes to the DB or its contents in a way that breaks the Taskcluster instance and requires ops intervention to restore the DB to a working state. 

Further clarification about what kinds of queries are to be supported, and to whom the postgres permissions are to be used to restrict their access, would help.

Reply:
The intent of the word "queries" here was read-only, but clarification would be good.  Also, I think we want to leave open the possibility of such read-only querying via views, rather than stored procedures.

In practice, the idea here is to give folks like Mike Trinkala access to the database for purposes of doing ETL-style analysis for cost monitoring and similar purposes.